pkg/datapath/linux/node_ids.go

```diff
@@ -73,10 +73,17 @@ func (n *linuxNodeHandler) getNodeIDForNode(node *nodeTypes.Node) uint16 {
 // been assigned. If any of the node IPs have an ID associated, then all other
 // node IPs receive the same. This might happen if we allocated a node ID from
 // the ipcache, where we don't have all node IPs but only one.
-func (n *linuxNodeHandler) allocateIDForNode(node *nodeTypes.Node) (uint16, error) {
+func (n *linuxNodeHandler) allocateIDForNode(oldNode *nodeTypes.Node, node *nodeTypes.Node) (uint16, error) {
+	var errs error
+
 	// Did we already allocate a node ID for any IP of that node?
 	nodeID := n.getNodeIDForNode(node)
 
+	var SPIChanged bool
+	if oldNode != nil {
+		SPIChanged = (oldNode.EncryptionKey != node.EncryptionKey)
+	}
```

Comment:
Hm, this defines whether we want to force an update of the node ID map because we know the SPI changed. Could it happen that we don't have an oldNode but the node ID map still contains an entry for that node IP? If so, we might want to default to true.

Maybe one such scenario is IP reuse after agent is down:
1. NodeA with IP_Z is created, mapping created.
2. Agent goes down.
3. NodeA is deleted.
4. Agent comes back.
5. SPI is incremented via a key rotation.
6. NodeB with IP_Z is created.

At 6 we want to update the existing entry in the node ID map, no?

Reply:
Yeah this is a really good point. The code I have here is using IP like a stable ID. 

I agree, changing it to default to true is probably simplest. We may get some spurious force updates to the map, but those should be minimal.

Reply:
This is kinda making me think of a bigger point. 

Do we cleanup eBPF Node map at all? For instance, what occurs in the same scenario above except NodeB uses a completely separate IP. Does NodeA's IP stay in the map indefinitely until something reuses it?

Reply:
Original issue fixed in latest push.